File: src/main/scala/torch_scala/examples/FourierNet.scala
package torch_scala.examples

import org.bytedeco.javacpp._
import org.bytedeco.javacpp.annotation._
import torch_scala.api.nn.Module


@Platform(include = Array("torch/all.h", "models/FourierNet.h", "<iostream>", "<vector>", "<map>"))
class FourierNet(val size: Int) extends Module {
  allocate(size)

  @native def allocate(@Cast(Array("int")) size: Int): Unit

  @StdVector @native private def train(@StdVector data: FloatPointer, steps: Int, @StdVector weights: FloatPointer): FloatPointer

  def train(data: Array[Float], steps: Int, weights: Array[Float]): Array[Float] = {
    val res = train(new FloatPointer(data:_*), steps, new FloatPointer(weights:_*))
    Array.range(0, data.length).map(i => res.get(i))
  }

  def loss(y1: Array[Float], y2: Array[Float], weights: Array[Float]): Float = {

    y1.zip(y2).map(p => p._1 - p._2).map(x => x*x).zip(weights).map(p => p._1 * p._2).sum

  }
}

File: src/main/java/torch_java/examples/FourierNet.java
package torch_java.examples;

import org.bytedeco.javacpp.*;
import org.bytedeco.javacpp.annotation.*;
import torch_java.api.nn.Module;

@Platform(
        include = {
                "torch/all.h",
                "models/FourierNet.h",
                "<iostream>",
                "<vector>",
                "<map>"
        })

@NoOffset public class FourierNet extends Module {


    public FourierNet(int size) {
        super();
        allocate(size);
    }

    public native void allocate(int size);

    // std::pair<float, std::vector<float> > train(std::vector<float> data, int steps, std::vector<float> weights)

    private native @StdVector FloatPointer train (@StdVector FloatPointer data, int steps, @StdVector FloatPointer weights);

    public float[] train(float[] data, int steps, float[] weights) {
        FloatPointer res = train(new FloatPointer(data), steps, new FloatPointer(weights));
        float[] res_array = new float[data.length];
        for(int i = 0; i < res_array.length; i++) {
            res_array[i] = res.get(i);
        }
        return res_array;
    }

    public float loss(float[] y1, float[] y2, float[] weights) {
        float res = 0;
        int n = y1.length;
        for(int i = 0; i < n; i++) {
            float dy = y1[i] - y2[i];
            res += dy * dy * weights[i];
        }
        return res;
    }

    public static void main(String[] args) throws InterruptedException {

        FourierNet net = new FourierNet(20);
        float[] pred = net.train(new float[]{1, 2, 3, 3, 4, 5, 6, 7, 6, 4, 3, 2, 4, 4, 5}, 500, new float[]{1,1,1,1,1,1,1,1,1,1,1,1,1,1,1});
        float loss = net.loss(pred, new float[]{1, 2, 3, 3, 4, 5, 6, 7, 6, 4, 3, 2, 4, 4, 5}, new float[]{1,1,1,1,1,1,1,1,1,1,1,1,1,1,1});
        System.out.println("loss = " + loss);
    }

}
